File: quick_sort/lib/quick_sort.rb
class QuickSort
  # Quick sort has average case time complexity O(nlogn), but worst
  # case O(n**2).

  # Not in-place. Uses O(n) memory.
  def self.sort1(array, &prc)
    prc ||= proc { |a, b| a <=> b }
    return array if array.length < 2

    pivot = array.first
    left = array[1..-1].select { |el| prc.call(el, pivot) == -1 }
    right = array[1..-1].select { |el| prc.call(el, pivot) != -1 }
    
    self.sort1(left, &prc) + [pivot] + self.sort1(right, &prc)
  end

  # In-place.
  def self.sort2!(array, start = 0, length = array.length, &prc)
    prc ||= proc { |a, b| a <=> b }

    if length > 1
      pivot_idx = self.partition(array, start, length, &prc)
      
      # because pivot_idx is rand, we have recursive calls made that cover the entire range of the array
      self.sort2!(array, start, pivot_idx - start, &prc)
      self.sort2!(array, pivot_idx + 1, length - 1 - pivot_idx, &prc)
    end 
    
    array
  end

  def self.partition(array, start, length, &prc)
    prc ||= proc { |a, b| a <=> b }

    pivot = array[start]
    pivot_idx = start

    # start + 1 because we onlly deal with the pivot val's position at the end of the partition
    (start + 1..start + length - 1).each do |idx|
      # if array[idx] val < pivot val, swap items and move barrier (pivot_idx) up one 
      if prc.call(pivot, array[idx]) == 1 
        array[pivot_idx + 1], array[idx] = array[idx], array[pivot_idx + 1]
        pivot_idx += 1
      end 
    end 

    # swap pivot directly to the left of where the barrier now is (pivot_idx) to put pivot at it's correct index 
    array[pivot_idx], array[start] = array[start], array[pivot_idx]
    pivot_idx
  end
end

File: 2022/quicksort.py
# avg time complexity O(nlog(n))
# space O(n)

def quick_sort(array):
    if len(array) <= 1:
        return array
    else:
        pivot = array[0]
        less = [x for x in array[1:] if x <= pivot]
        greater = [x for x in array[1:] if x > pivot]
        return quick_sort(less) + [pivot] + quick_sort(greater)

# avg time O(nlogn)
# avg space O(log(n))
def quick_sort_in_place(array, start, end):
    if start >= end:
        return
    pivot_index = partition(array, start, end)
    quick_sort_in_place(array, start, pivot_index - 1)
    quick_sort_in_place(array, pivot_index + 1, end)

def partition(array, start, end):
    pivot = array[end]
    pivot_index = start
    for i in range(start, end):
        if array[i] <= pivot:
            array[i], array[pivot_index] = array[pivot_index], array[i]
            pivot_index += 1
    array[pivot_index], array[end] = array[end], array[pivot_index]
    return pivot_index